server/src/main/java/io/druid/curator/discovery/CuratorDruidNodeDiscoveryProvider.java

```diff
@@ -0,0 +1,327 @@
+      nodeTypeWatcher = new NodeTypeWatcher(
+          listenerExecutor,
+          curatorFramework,
+          config.getInternalDiscoveryPath(),
+          jsonMapper,
+          nodeType
+      );
+      nodeTypeWatcher.start();
+      nodeTypeWatchers.put(nodeType, nodeTypeWatcher);
+      log.info("Created/Started NodeTypeWatcher for nodeType [%s].", nodeType);
+    }
+
+    return nodeTypeWatchers.get(nodeType);
+  }
+
+  @LifecycleStart
+  public synchronized void start()
+  {
+    log.info("started");
+  }
+
+  @LifecycleStop
+  public synchronized void stop()
+  {
+    if (stopped) {
+      return;
+    }
+
+    log.info("stopping");
+
+    stopped = true;
+
+    for (NodeTypeWatcher watcher : nodeTypeWatchers.values()) {
+      watcher.stop();
+    }
+    listenerExecutor.shutdownNow();
+
+    log.info("stopped");
+  }
+
+  private static class NodeTypeWatcher implements DruidNodeDiscovery, PathChildrenCacheListener
+  {
+    private static final Logger log = new Logger(NodeTypeWatcher.class);
+
+    private final CuratorFramework curatorFramework;
+
+    private final String nodeType;
+    private final ObjectMapper jsonMapper;
+
+    // hostAndPort -> DiscoveryDruidNode
+    private final Map<String, DiscoveryDruidNode> nodes = new ConcurrentHashMap<>();
+
+    private final PathChildrenCache cache;
+    private final ExecutorService cacheExecutor;
+
+    private final ExecutorService listenerExecutor;
+
+    private final List<DruidNodeDiscovery.Listener> nodeListeners = new ArrayList();
+
+    NodeTypeWatcher(
+        ExecutorService listenerExecutor,
+        CuratorFramework curatorFramework,
+        String basePath,
+        ObjectMapper jsonMapper,
+        String nodeType
+    )
+    {
+      this.listenerExecutor = listenerExecutor;
+      this.curatorFramework = curatorFramework;
+      this.nodeType = nodeType;
+      this.jsonMapper = jsonMapper;
+
+      // This is required to be single threaded from Docs in PathChildrenCache;
+      this.cacheExecutor = Execs.singleThreaded(String.format("NodeTypeWatcher[%s]-[%%d]", nodeType));
+      this.cache = new PathChildrenCache(
+          curatorFramework,
+          ZKPaths.makePath(basePath, nodeType),
+          true,
+          true,
+          cacheExecutor
+      );
+    }
+
+    @Override
+    public Set<DiscoveryDruidNode> getAllNodes()
+    {
+      return new ImmutableSet.Builder().addAll(nodes.values()).build();
+    }
+
+    @Override
+    public synchronized void registerListener(DruidNodeDiscovery.Listener listener)
+    {
+      for (DiscoveryDruidNode node : nodes.values()) {
+        listenerExecutor.submit(() -> {
+          try {
+            listener.nodeAdded(node);
+          }
+          catch (Exception ex) {
+            log.error(ex, "Exception occured in DiscoveryDruidNode.nodeAdded(node=[%s]) in listener [%s].", node, listener);
+          }
+        });
+      }
+
+      nodeListeners.add(listener);
```

Comment:
I think it might be better to add to the list first and then register the added stuff.